skl2onnx/operator_converters/nearest_neighbours.py

```diff
@@ -5,498 +5,236 @@
-
-            container.add_node('Equal', [labels_name[i], topk_labels_name],
-                               output_label_name[i],
-                               name=scope.get_unique_operator_name('Equal'))
-            apply_cast(scope, output_label_name[i], output_cast_label_name[i],
-                       container, to=onnx_proto.TensorProto.FLOAT)
-            apply_mul(scope, [output_cast_label_name[i], weights_val],
-                      weighted_distance_name, container, broadcast=0)
-            container.add_node('ReduceSum', weighted_distance_name,
-                               output_label_reduced_name[i], axes=[1],
-                               name=scope.get_unique_operator_name(
-                                   'ReduceSum'))
+    X = operator.inputs[0]
+    op = operator.raw_operator
+    opv = container.target_opset
+    dtype = container.dtype
+
+    if X.type.__class__ == Int64TensorType:
+        X = OnnxCast(X, to=container.proto_dtype, op_version=opv)
+
+    options = container.get_options(op, dict(optim=None))
+
+    single_reg = (not hasattr(op, '_y') or len(op._y.shape) == 1 or
+                  len(op._y.shape) == 2 and op._y.shape[1] == 1)
+    ndim = 1 if single_reg else op._y.shape[1]
+
+    metric = op.effective_metric_
+    neighb = op._fit_X.astype(container.dtype)
+    k = op.n_neighbors
+    training_labels = op._y if hasattr(op, '_y') else None
+    distance_kwargs = {}
+    if metric == 'minkowski':
+        if op.p != 2:
+            distance_kwargs['p'] = op.p
+        else:
+            metric = "euclidean"
+
+    weights = op.weights if hasattr(op, 'weights') else 'distance'
+    if weights == 'uniform':
+        top_indices = onnx_nearest_neighbors_indices(
+            X, neighb, k, metric=metric, dtype=dtype,
+            op_version=opv, optim=options.get('optim', None),
+            **distance_kwargs)
+        top_distances = None
+    elif weights == 'distance':
+        top_indices, top_distances = onnx_nearest_neighbors_indices(
+            X, neighb, k, metric=metric, dtype=dtype,
+            op_version=opv, keep_distances=True,
+            optim=options.get('optim', None),
+            **distance_kwargs)
+    else:
+        raise RuntimeError(
+            "Unable to convert KNeighborsRegressor when weights is callable.")
+
+    shape = OnnxShape(top_indices, op_version=opv)
+    flattened = OnnxFlatten(top_indices, op_version=opv)
+    if training_labels is not None:
+        if ndim > 1:
+            # shape = (ntargets, ) + shape
+            training_labels = training_labels.T
+            shape = OnnxConcat(np.array([ndim], dtype=np.int64),
+                               shape, op_version=opv, axis=0)
+            axis = 2
+        else:
+            training_labels = training_labels.ravel()
+            axis = 1
+
+        if training_labels.dtype == np.int32:
+            training_labels = training_labels.astype(np.int64)
+        extracted = OnnxArrayFeatureExtractor(
```

Comment:
Is OnnxArrayFeatureExtractor different from ArrayFeatureExtractor? The reason why KNN converter couldn't handle batch prediction was ArrayFeatureExtractor as they can only handle one homogenous indices(Y is a tensor of int), which means you can't extract different values for different test example(based on their nearest neightbour).

Reply:
That's the same. It is used to extract the neighbors labels. It is equivalent to GatherElements if the target dimension is one but still needed if the target dimension is more than one. The previous version had more than one ArrayFeatureExtractor.